Remotion/Core/Development.Analyzers/RMCORE0001_SafeContextAnalyzer.cs

```diff
@@ -0,0 +1,258 @@
+        return;
+
+      var memberAccessOriginalDefinitionSymbol = memberAccessSymbolInfo.Symbol.OriginalDefinition;
+
+      var match = eventSymbolAndDiagnostics.FirstOrDefault(m => SymbolEqualityComparer.Default.Equals(m.eventSymbol, memberAccessOriginalDefinitionSymbol));
+      if(match != default)
+        match.CreateDiagnostic.Invoke(ctx, match.eventSymbol);
+    }
+
+    private static void AnalyzeInvocationExpression (SyntaxNodeAnalysisContext ctx, IReadOnlyCollection<(IMethodSymbol methodSymbol, Action<SyntaxNodeAnalysisContext, ISymbol> CreateDiagnostic)> methodSymbols)
+    {
+      var invocationSymbolInfo = ctx.SemanticModel.GetSymbolInfo(ctx.Node);
+
+      if (invocationSymbolInfo.Symbol == null)
+        return;
+
+      var invocationOriginalDefinitionSymbol = invocationSymbolInfo.Symbol.OriginalDefinition;
+
+      var match = methodSymbols.FirstOrDefault(m => SymbolEqualityComparer.Default.Equals(m.methodSymbol, invocationOriginalDefinitionSymbol));
+      if (match != default)
+        match.CreateDiagnostic.Invoke(ctx, match.methodSymbol);
+    }
+
+    private static void AnalyzeObjectCreationExpression (SyntaxNodeAnalysisContext ctx, IReadOnlyCollection<(IMethodSymbol methodSymbol, Action<SyntaxNodeAnalysisContext, ISymbol> CreateDiagnostic)> ctorMethodsWithDiagnostics)
+    {
+      var objectCreationInfo = ctx.SemanticModel.GetSymbolInfo(ctx.Node);
+
+      if (objectCreationInfo.Symbol == null)
+        return;
+
+      var objectCreationInfoOriginalDefinitionSymbol = (IMethodSymbol)objectCreationInfo.Symbol.OriginalDefinition;
+
+      var match = ctorMethodsWithDiagnostics.FirstOrDefault(m => SymbolEqualityComparer.Default.Equals(m.Item1, objectCreationInfoOriginalDefinitionSymbol));
+      if (match != default)
+        match.CreateDiagnostic.Invoke(ctx, match.methodSymbol);
+    }
+
+    private static IEnumerable<IMethodSymbol> GetMethodSymbols (INamedTypeSymbol? typeSymbol, string name)
+    {
+      if (typeSymbol == null)
+        return Enumerable.Empty<IMethodSymbol>();
+
+      return typeSymbol
+          .GetMembers(name)
+          .Cast<IMethodSymbol>()
+          .ToList();
+    }
+
+    private static IEnumerable<IMethodSymbol> GetCtorSymbols (INamedTypeSymbol? typeSymbol)
+    {
+      if (typeSymbol == null)
+        return Enumerable.Empty<IMethodSymbol>();
+
+      return typeSymbol.Constructors.ToList();
+    }
+
+    private static IEnumerable<IMethodSymbol> GetAddEventMethodSymbol (INamedTypeSymbol? typeSymbol, string name)
+    {
+      if (typeSymbol == null)
+        return Enumerable.Empty<IMethodSymbol>();
+
+
+      return typeSymbol
+          .GetMembers(name)
+          .Cast<IEventSymbol>()
+          .Where(s => s.AddMethod != null)
+          .Select(s => s.AddMethod!);
+    }
+
+    private static void AddRange (List<(IMethodSymbol, Action<SyntaxNodeAnalysisContext, ISymbol>)> listWithBoth, IEnumerable<IMethodSymbol> symbols, Action<SyntaxNodeAnalysisContext, ISymbol> action)
```

Comment:
We need a better name for this. We might also pull in more of the logic to allow the main registration above to be a bit more streamlined. Ideally, one could add new methods just by adding a few lines in `RegisterCompilationStartAction` action above.